Source/Core/Core/AchievementManager.cpp

```diff
@@ -111,4 +123,210 @@ AchievementManager::ResponseType AchievementManager::Request(
+{
+  if (!m_dll_enabled)
+    return;
+  RA_DoAchievementsFrame();
+}
+
+bool WideStringToUTF8String(std::string& dest, const std::wstring_view& str)
+{
+  int mblen = WideCharToMultiByte(CP_UTF8, 0, str.data(), static_cast<int>(str.length()), nullptr,
+                                  0, nullptr, nullptr);
+  if (mblen < 0)
+    return false;
+
+  dest.resize(mblen);
+  if (mblen > 0 && WideCharToMultiByte(CP_UTF8, 0, str.data(), static_cast<int>(str.length()),
+                                       dest.data(), mblen, nullptr, nullptr) < 0)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+std::string WideStringToUTF8String(const std::wstring_view& str)
+{
+  std::string ret;
+  if (!WideStringToUTF8String(ret, str))
+    ret.clear();
+
+  return ret;
+}
+
+std::vector<std::tuple<int, std::string, bool>> AchievementManager::GetMenuItems()
+{
+  if (!m_dll_enabled)
+    return std::vector<std::tuple<int, std::string, bool>>();
+  std::array<RA_MenuItem, 64> items;
+  const int num_items = RA_GetPopupMenuItems(items.data());
+
+  std::vector<std::tuple<int, std::string, bool>> ret;
+  ret.reserve(static_cast<u32>(num_items));
+
+  for (int i = 0; i < num_items; i++)
+  {
+    const RA_MenuItem& it = items[i];
+    if (!it.sLabel)
+    {
+      // separator
+      ret.emplace_back(0, std::string(), false);
+    }
+    else
+    {
+      // option, maybe checkable
+//      ret.emplace_back(static_cast<int>(it.nID), it.sLabel, it.bChecked);
+      ret.emplace_back(static_cast<int>(it.nID), WideStringToUTF8String(it.sLabel), it.bChecked);
+    }
+  }
+
+  return ret;
+}
+
+void AchievementManager::ActivateMenuItem(int item)
+{
+  if (!m_dll_enabled)
+    return;
+  RA_InvokeDialog(item);
+}
+
+int AchievementManager::RACallbackIsActive()
+{
+  return m_game_id;
+}
+
+void AchievementManager::RACallbackCauseUnpause()
+{
+  if (Core::GetState() != Core::State::Uninitialized)
+    Core::SetState(Core::State::Running);
+}
+
+void AchievementManager::RACallbackCausePause()
+{
+  if (Core::GetState() != Core::State::Uninitialized)
+    Core::SetState(Core::State::Paused);
+}
+
+void AchievementManager::RACallbackRebuildMenu()
+{
+  // unused
+}
+
+void AchievementManager::RACallbackEstimateTitle(char* buf)
+{
+  strcpy(buf, m_filename.c_str());
+}
+
+void AchievementManager::RACallbackResetEmulator()
+{
+  Core::Stop();
+}
+
+void AchievementManager::RACallbackLoadROM(const char* unused)
+{
+  // unused
+}
+
+unsigned char AchievementManager::RACallbackReadMemory(unsigned int address)
+{
+  return m_memory_manager->Read_U8(address);
+}
+
+unsigned int AchievementManager::RACallbackReadBlock(unsigned int address, unsigned char* buffer,
+                                                     unsigned int bytes)
+{
+  m_memory_manager->CopyFromEmu(buffer, address, bytes);
+  return bytes;
+}
+
+void AchievementManager::RACallbackWriteMemory(unsigned int address, unsigned char value)
+{
+  m_memory_manager->Write_U8(value, address);
+}
```

Comment:
These are not the functions external tools should be using to read and write memory, please use the `Host` stuff in `MMU.h` instead.

Reply:
Even then, that might not be appropriate, as these callbacks can occur on a separate thread created by RAIntegration itself (either with some worker UI thread or the rich presence thread). I got caught by that detail when creating BizHawk's integration, resulting in rather annoying workarounds [due to our peeks not being thread safe](https://github.com/TASEmulators/BizHawk/blob/fbe1f9b53e0b319b6a9b8405cb991a1bd4dafe23/src/BizHawk.Client.EmuHawk/RetroAchievements/RetroAchievements.Memory.cs#L14-L94). Even then, given the "memory maps" that got decided in rcheevos, it would probably be better to directly peek/poke at ram using `m_pRAM`/`m_pEXRAM` in Memmap.cpp, which then threading details probably don't matter.

Reply:
Yeah, I mentioned this over in the Discord, but given the limitations RAIntegration has with memory maps I don't really see any good way to handle this other than giving them the physical memory directly. Which is still sucks, by the way, because games address memory by virtual addresses and not physical addresses, so every achievement-maker will have to keep a weird map in their head. It's also going to make achievements for games that have their pages mapped in a different order virtually vs physically a pain in the ass to deal with. I don't like it.